// @hubspot/local-dev-lib/http/getAxiosConfig.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAxiosConfig = exports.DEFAULT_USER_AGENT_HEADERS = void 0;
const package_json_1 = require("../package.json");
const config_1 = require("../config");
const urls_1 = require("../lib/urls");
exports.DEFAULT_USER_AGENT_HEADERS = {
    'User-Agent': `HubSpot Local Dev Lib/${package_json_1.version}`,
};
const DEFAULT_TRANSITIONAL = {
    clarifyTimeoutError: true,
};
function getAxiosConfig(options) {
    const { env, localHostOverride, headers, ...rest } = options;
    const { httpTimeout, httpUseLocalhost } = (0, config_1.getAndLoadConfigIfNeeded)();
    return {
        baseURL: (0, urls_1.getHubSpotApiOrigin)(env, localHostOverride ? false : httpUseLocalhost),
        headers: {
            ...exports.DEFAULT_USER_AGENT_HEADERS,
            ...(headers || {}),
        },
        timeout: httpTimeout || 15000,
        transitional: DEFAULT_TRANSITIONAL,
        ...rest,
    };
}
exports.getAxiosConfig = getAxiosConfig;
