// @hubspot/local-dev-lib/api/customObjects.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.deleteObjectSchema = exports.fetchObjectSchemas = exports.fetchObjectSchema = exports.updateObjectSchema = exports.createObjectSchema = exports.batchCreateObjects = void 0;
const http_1 = __importDefault(require("../http"));
const CUSTOM_OBJECTS_API_PATH = 'crm/v3/objects';
const SCHEMA_API_PATH = 'crm-object-schemas/v3/schemas';
async function batchCreateObjects(accountId, objectTypeId, objects) {
    return http_1.default.post(accountId, {
        url: `${CUSTOM_OBJECTS_API_PATH}/${objectTypeId}/batch/create`,
        data: objects,
    });
}
exports.batchCreateObjects = batchCreateObjects;
async function createObjectSchema(accountId, schema) {
    return http_1.default.post(accountId, {
        url: SCHEMA_API_PATH,
        data: schema,
    });
}
exports.createObjectSchema = createObjectSchema;
async function updateObjectSchema(accountId, schemaObjectType, schema) {
    return http_1.default.patch(accountId, {
        url: `${SCHEMA_API_PATH}/${schemaObjectType}`,
        data: schema,
    });
}
exports.updateObjectSchema = updateObjectSchema;
async function fetchObjectSchema(accountId, schemaObjectType) {
    return http_1.default.get(accountId, {
        url: `${SCHEMA_API_PATH}/${schemaObjectType}`,
    });
}
exports.fetchObjectSchema = fetchObjectSchema;
async function fetchObjectSchemas(accountId) {
    return http_1.default.get(accountId, {
        url: SCHEMA_API_PATH,
    });
}
exports.fetchObjectSchemas = fetchObjectSchemas;
async function deleteObjectSchema(accountId, schemaObjectType) {
    return http_1.default.delete(accountId, {
        url: `${SCHEMA_API_PATH}/${schemaObjectType}`,
    });
}
exports.deleteObjectSchema = deleteObjectSchema;
