// @hubspot/local-dev-lib/api/hubdb.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.deleteRows = exports.fetchRows = exports.createRows = exports.deleteTable = exports.publishTable = exports.updateTable = exports.createTable = exports.fetchTable = void 0;
const http_1 = __importDefault(require("../http"));
const HUBDB_API_PATH = 'cms/v3/hubdb';
async function fetchTable(accountId, tableId) {
    return http_1.default.get(accountId, {
        url: `${HUBDB_API_PATH}/tables/${tableId}`,
    });
}
exports.fetchTable = fetchTable;
async function createTable(accountId, schema) {
    return http_1.default.post(accountId, {
        url: `${HUBDB_API_PATH}/tables`,
        data: schema,
    });
}
exports.createTable = createTable;
async function updateTable(accountId, tableId, schema) {
    return http_1.default.patch(accountId, {
        url: `${HUBDB_API_PATH}/tables/${tableId}/draft`,
        data: schema,
    });
}
exports.updateTable = updateTable;
async function publishTable(accountId, tableId) {
    return http_1.default.post(accountId, {
        url: `${HUBDB_API_PATH}/tables/${tableId}/draft/publish`,
        headers: {
            'Content-Type': 'application/json',
        },
    });
}
exports.publishTable = publishTable;
async function deleteTable(accountId, tableId) {
    return http_1.default.delete(accountId, {
        url: `${HUBDB_API_PATH}/tables/${tableId}`,
    });
}
exports.deleteTable = deleteTable;
async function createRows(accountId, tableId, rows) {
    return http_1.default.post(accountId, {
        url: `${HUBDB_API_PATH}/tables/${tableId}/rows/draft/batch/create`,
        data: { inputs: rows },
    });
}
exports.createRows = createRows;
async function fetchRows(accountId, tableId, params = {}) {
    return http_1.default.get(accountId, {
        url: `${HUBDB_API_PATH}/tables/${tableId}/rows/draft`,
        params,
    });
}
exports.fetchRows = fetchRows;
async function deleteRows(accountId, tableId, rowIds) {
    return http_1.default.post(accountId, {
        url: `${HUBDB_API_PATH}/tables/${tableId}/rows/draft/batch/purge`,
        data: { inputs: rowIds },
    });
}
exports.deleteRows = deleteRows;
