// @hubspot/local-dev-lib/api/localDevAuth.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.fetchScopeData = exports.fetchAccessToken = void 0;
const getAxiosConfig_1 = require("../http/getAxiosConfig");
const http_1 = __importDefault(require("../http"));
const environments_1 = require("../constants/environments");
const axios_1 = __importDefault(require("axios"));
const LOCALDEVAUTH_API_AUTH_PATH = 'localdevauth/v1/auth';
async function fetchAccessToken(personalAccessKey, env = environments_1.ENVIRONMENTS.PROD, portalId) {
    const axiosConfig = (0, getAxiosConfig_1.getAxiosConfig)({
        env,
        localHostOverride: true,
        url: `${LOCALDEVAUTH_API_AUTH_PATH}/refresh`,
        data: {
            encodedOAuthRefreshToken: personalAccessKey,
        },
        params: portalId ? { portalId } : {},
    });
    const { data } = await (0, axios_1.default)({
        ...axiosConfig,
        method: 'post',
    });
    return data;
}
exports.fetchAccessToken = fetchAccessToken;
async function fetchScopeData(accountId, scopeGroup) {
    return http_1.default.get(accountId, {
        url: `localdevauth/v1/auth/check-scopes`,
        params: { scopeGroup },
    });
}
exports.fetchScopeData = fetchScopeData;
