// @hubspot/local-dev-lib/api/projects.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.cancelStagedBuild = exports.deleteFileFromBuild = exports.uploadFileToBuild = exports.queueBuild = exports.provisionBuild = exports.fetchDeployComponentsMetadata = exports.fetchProjectSettings = exports.getDeployStructure = exports.getDeployStatus = exports.deployProject = exports.getBuildStructure = exports.getBuildStatus = exports.fetchProjectBuilds = exports.fetchPlatformVersions = exports.deleteProject = exports.downloadProject = exports.fetchProject = exports.uploadProject = exports.createProject = exports.fetchProjects = void 0;
const http_1 = __importDefault(require("../http"));
const fs_1 = __importDefault(require("fs"));
const PROJECTS_API_PATH = 'dfs/v1/projects';
const PROJECTS_DEPLOY_API_PATH = 'dfs/deploy/v1';
const DEVELOPER_PROJECTS_API_PATH = 'developer/projects/v1';
async function fetchProjects(accountId) {
    return http_1.default.get(accountId, {
        url: PROJECTS_API_PATH,
    });
}
exports.fetchProjects = fetchProjects;
async function createProject(accountId, name) {
    return http_1.default.post(accountId, {
        url: PROJECTS_API_PATH,
        data: {
            name,
        },
    });
}
exports.createProject = createProject;
async function uploadProject(accountId, projectName, projectFile, uploadMessage, platformVersion) {
    const formData = {
        file: fs_1.default.createReadStream(projectFile),
        uploadMessage,
    };
    if (platformVersion) {
        formData.platformVersion = platformVersion;
    }
    return http_1.default.post(accountId, {
        url: `${PROJECTS_API_PATH}/upload/${encodeURIComponent(projectName)}`,
        timeout: 60000,
        data: formData,
        headers: { 'Content-Type': 'multipart/form-data' },
    });
}
exports.uploadProject = uploadProject;
async function fetchProject(accountId, projectName) {
    return http_1.default.get(accountId, {
        url: `${PROJECTS_API_PATH}/${encodeURIComponent(projectName)}`,
    });
}
exports.fetchProject = fetchProject;
async function downloadProject(accountId, projectName, buildId) {
    return http_1.default.get(accountId, {
        url: `${PROJECTS_API_PATH}/${encodeURIComponent(projectName)}/builds/${buildId}/archive-full`,
        responseType: 'arraybuffer',
        headers: { accept: 'application/zip', 'Content-Type': 'application/json' },
    });
}
exports.downloadProject = downloadProject;
async function deleteProject(accountId, projectName) {
    return http_1.default.delete(accountId, {
        url: `${PROJECTS_API_PATH}/${encodeURIComponent(projectName)}`,
    });
}
exports.deleteProject = deleteProject;
async function fetchPlatformVersions(accountId) {
    return http_1.default.get(accountId, {
        url: `${DEVELOPER_PROJECTS_API_PATH}/platformVersion`,
    });
}
exports.fetchPlatformVersions = fetchPlatformVersions;
async function fetchProjectBuilds(accountId, projectName, params = {}) {
    return http_1.default.get(accountId, {
        url: `${PROJECTS_API_PATH}/${encodeURIComponent(projectName)}/builds`,
        params,
    });
}
exports.fetchProjectBuilds = fetchProjectBuilds;
async function getBuildStatus(accountId, projectName, buildId) {
    return http_1.default.get(accountId, {
        url: `${PROJECTS_API_PATH}/${encodeURIComponent(projectName)}/builds/${buildId}/status`,
    });
}
exports.getBuildStatus = getBuildStatus;
async function getBuildStructure(accountId, projectName, buildId) {
    return http_1.default.get(accountId, {
        url: `dfs/v1/builds/by-project-name/${encodeURIComponent(projectName)}/builds/${buildId}/structure`,
    });
}
exports.getBuildStructure = getBuildStructure;
async function deployProject(accountId, projectName, buildId) {
    return http_1.default.post(accountId, {
        url: `${PROJECTS_DEPLOY_API_PATH}/deploys/queue/async`,
        data: {
            projectName,
            buildId,
        },
    });
}
exports.deployProject = deployProject;
async function getDeployStatus(accountId, projectName, deployId) {
    return http_1.default.get(accountId, {
        url: `${PROJECTS_DEPLOY_API_PATH}/deploy-status/projects/${encodeURIComponent(projectName)}/deploys/${deployId}`,
    });
}
exports.getDeployStatus = getDeployStatus;
async function getDeployStructure(accountId, projectName, deployId) {
    return http_1.default.get(accountId, {
        url: `${PROJECTS_DEPLOY_API_PATH}/deploys/by-project-name/${encodeURIComponent(projectName)}/deploys/${deployId}/structure`,
    });
}
exports.getDeployStructure = getDeployStructure;
async function fetchProjectSettings(accountId, projectName) {
    return http_1.default.get(accountId, {
        url: `${PROJECTS_API_PATH}/${encodeURIComponent(projectName)}/settings`,
    });
}
exports.fetchProjectSettings = fetchProjectSettings;
async function fetchDeployComponentsMetadata(accountId, projectId) {
    return http_1.default.get(accountId, {
        url: `${PROJECTS_API_PATH}/by-id/${projectId}/deploy-components-metadata`,
    });
}
exports.fetchDeployComponentsMetadata = fetchDeployComponentsMetadata;
async function provisionBuild(accountId, projectName, platformVersion) {
    return http_1.default.post(accountId, {
        url: `${PROJECTS_API_PATH}/${encodeURIComponent(projectName)}/builds/staged/provision`,
        params: { platformVersion },
        headers: { 'Content-Type': 'application/json' },
        timeout: 50000,
    });
}
exports.provisionBuild = provisionBuild;
async function queueBuild(accountId, projectName, platformVersion) {
    return http_1.default.post(accountId, {
        url: `${PROJECTS_API_PATH}/${encodeURIComponent(projectName)}/builds/staged/queue`,
        params: { platformVersion },
        headers: { 'Content-Type': 'application/json' },
    });
}
exports.queueBuild = queueBuild;
async function uploadFileToBuild(accountId, projectName, filePath, path) {
    return http_1.default.put(accountId, {
        url: `${PROJECTS_API_PATH}/${encodeURIComponent(projectName)}/builds/staged/files/${encodeURIComponent(path)}`,
        data: {
            file: fs_1.default.createReadStream(filePath),
        },
        headers: { 'Content-Type': 'multipart/form-data' },
    });
}
exports.uploadFileToBuild = uploadFileToBuild;
async function deleteFileFromBuild(accountId, projectName, path) {
    return http_1.default.delete(accountId, {
        url: `${PROJECTS_API_PATH}/${encodeURIComponent(projectName)}/builds/staged/files/${encodeURIComponent(path)}`,
    });
}
exports.deleteFileFromBuild = deleteFileFromBuild;
async function cancelStagedBuild(accountId, projectName) {
    return http_1.default.post(accountId, {
        url: `${PROJECTS_API_PATH}/${encodeURIComponent(projectName)}/builds/staged/cancel`,
        headers: { 'Content-Type': 'application/json' },
    });
}
exports.cancelStagedBuild = cancelStagedBuild;
