// @hubspot/local-dev-lib/api/sandboxSync.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.fetchTypes = exports.fetchTaskStatus = exports.initiateSync = void 0;
const http_1 = __importDefault(require("../http"));
const api_1 = require("../constants/api");
const SANDBOXES_SYNC_API_PATH = 'sandboxes-sync/v1';
async function initiateSync(fromHubId, toHubId, tasks, sandboxHubId) {
    return http_1.default.post(fromHubId, {
        data: {
            command: 'SYNC',
            fromHubId,
            toHubId,
            sandboxHubId,
            tasks,
        },
        timeout: api_1.SANDBOX_TIMEOUT,
        url: `${SANDBOXES_SYNC_API_PATH}/tasks/initiate/async`,
    });
}
exports.initiateSync = initiateSync;
async function fetchTaskStatus(accountId, taskId) {
    return http_1.default.get(accountId, {
        url: `${SANDBOXES_SYNC_API_PATH}/tasks/${taskId}`,
    });
}
exports.fetchTaskStatus = fetchTaskStatus;
async function fetchTypes(accountId, toHubId) {
    return http_1.default.get(accountId, {
        url: `${SANDBOXES_SYNC_API_PATH}/types${toHubId ? `?toHubId=${toHubId}` : ''}`,
    });
}
exports.fetchTypes = fetchTypes;
