// @hubspot/local-dev-lib/api/secrets.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.fetchSecrets = exports.deleteSecret = exports.updateSecret = exports.addSecret = void 0;
const http_1 = __importDefault(require("../http"));
const SECRETS_API_PATH = 'cms/v3/functions/secrets';
async function addSecret(accountId, key, value) {
    return http_1.default.post(accountId, {
        url: SECRETS_API_PATH,
        data: {
            key,
            secret: value,
        },
    });
}
exports.addSecret = addSecret;
async function updateSecret(accountId, key, value) {
    return http_1.default.put(accountId, {
        url: SECRETS_API_PATH,
        data: {
            key,
            secret: value,
        },
    });
}
exports.updateSecret = updateSecret;
async function deleteSecret(accountId, key) {
    return http_1.default.delete(accountId, {
        url: `${SECRETS_API_PATH}/${key}`,
    });
}
exports.deleteSecret = deleteSecret;
async function fetchSecrets(accountId) {
    return http_1.default.get(accountId, {
        url: `${SECRETS_API_PATH}`,
    });
}
exports.fetchSecrets = fetchSecrets;
