// @hubspot/ui-extensions-dev-server/dist/lib/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PLATFORM_VERSION = exports.SERVER_CAPABILITIES = exports.WEBSOCKET_MESSAGE_VERSION = exports.EXTENSIONS_MESSAGE_VERSION = exports.ROLLUP_OPTIONS = exports.WEBSOCKET_DEFAULT_PORT = exports.EXPRESS_DEFAULT_PORT = exports.VITE_DEV_SERVER_ID = exports.EXPRESS_SERVER_ID = exports.MANIFEST_FILE = exports.MAIN_APP_CONFIG = exports.OUTPUT_DIR = void 0;
exports.OUTPUT_DIR = 'dist';
exports.MAIN_APP_CONFIG = 'app.json';
exports.MANIFEST_FILE = 'manifest.json';
exports.EXPRESS_SERVER_ID = 'ui-extensions-dev-server';
exports.VITE_DEV_SERVER_ID = 'ui-extensions-vite-dev-server';
exports.EXPRESS_DEFAULT_PORT = 5173;
exports.WEBSOCKET_DEFAULT_PORT = 5174;
exports.ROLLUP_OPTIONS = {
    // Deps to exclude from the bundle
    external: ['react', 'react-dom', '@remote-ui/react'],
    output: {
        // Maps libs to the variables to be injected via the window
        globals: {
            react: 'React',
            '@remote-ui/react': 'RemoteUI',
        },
        extend: true,
    },
};
exports.EXTENSIONS_MESSAGE_VERSION = 2;
exports.WEBSOCKET_MESSAGE_VERSION = 1;
exports.SERVER_CAPABILITIES = [
    // Supports running app functions locally
    'app-functions-local-dev',
    'source-id',
    'account-id',
];
exports.PLATFORM_VERSION = {
    V20231: '2023.1',
    V20232: '2023.2',
};
