// @hubspot/local-dev-lib/lib/logging/table.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getTableHeader = exports.getTableContents = void 0;
const chalk_1 = __importDefault(require("chalk"));
const table_1 = require("table");
const objectUtils_1 = require("../../utils/objectUtils");
const tableConfigDefaults = {
    singleLine: true,
    border: {
        topBody: '',
        topJoin: '',
        topLeft: '',
        topRight: '',
        bottomBody: '',
        bottomJoin: '',
        bottomLeft: '',
        bottomRight: '',
        bodyLeft: '',
        bodyRight: '',
        bodyJoin: '',
        joinBody: '',
        joinLeft: '',
        joinRight: '',
        joinJoin: '',
    },
    columnDefault: {
        paddingLeft: 0,
        paddingRight: 1,
    },
    drawHorizontalLine: () => {
        return false;
    },
};
function getTableContents(
// eslint-disable-next-line @typescript-eslint/no-explicit-any
tableData = [], tableConfig = {}) {
    const mergedConfig = (0, objectUtils_1.mergeDeep)({}, tableConfigDefaults, tableConfig);
    return (0, table_1.table)(tableData, mergedConfig);
}
exports.getTableContents = getTableContents;
function getTableHeader(headerItems) {
    return headerItems.map(headerItem => chalk_1.default.bold(headerItem));
}
exports.getTableHeader = getTableHeader;
